Open issue popup on map load

gileri opened the pull request:
Following [this comment](https://github.com/osm-fr/osmose-frontend/issues/155#issuecomment-517908455) on #155.

I added a parameter in the RSS feeds and error lists to allow opening the choosen issue popup on the map from RSS feeds (item link), error lists (`pos` column), GPX etc.

Currently it opens based on the center coordinates of the viewport. It works well when coming from generated links as the viewport is centered on the issue. However if the user pans the map and share its link the popup will not match the error coords, but the center of the viewport. If anyone find how to do so feel free to push to my branch.

But even with this slight bug it improves the editing workflow (it helps a lot in my case) so I think it could be merged.

gileri commented on web/static/map/map.js:
```diff
@@ -37,6 +37,7 @@ export function initMap() {
   urlVars.level = urlVars.level || Cookies.get('last_level') || '1';
   urlVars.tags = urlVars.tags || Cookies.get('last_tags');
   urlVars.fixable = urlVars.fixable || Cookies.get('last_fixable');
+  urlVars.errorId = urlVars.errorId || Cookies.get('error_id');
```
Do the `errorId` URL parameter name seem correct ? As it will be used on RSS feeds and other "persistent" resources I think it must be choosen wisely.

frodrigo replied:
Should be `marker_uuid`

gileri replied:
Fixed in 1cd2f26f94ab6a6c3092f6cc6e84c6145d01da3b.

frodrigo commented on web/views/errors/list.rss:
```diff
@@ -61,7 +61,7 @@
     </description>
     <category>{{res["item"]}}</category>
-%    url = 'http://%s/%s/map/#%s&zoom=13&lat=%s&lon=%s&level=%s&tags=&fixable=' % (website, lang, query, lat, lon, res["level"])
+%    url = 'http://%s/%s/map/#zoom=16&lat=%s&lon=%s&level=%s&tags=&fixable=&errorId=%s' % (website, lang, lat, lon, res["level"], res["uuid"])
```
Need to remove `query` ?

gileri replied:
Good catch ! I removed it temporarily as it resulted in two `zoom` fragment parameters being present : one from query, one hardcoded. Restored it in 6efcd36.

That raises the following question : which zoom should be used in those kind of links ?

- The one from the user at the risk that the popup couldn't be opened ?
- The hardcoded high zoom to allow issues to be shown in dense areas ?

There is a third solution that seem a bit too complicated to me : pass the `marker_uuid` to the API through the Marker VectorGrid so that the API will always return it even if the query `LIMIT` is reached (with something like `ORDER BY marker.uuid = marker_uuid` DESC).

gileri replied:
I don't think that's blocking for this PR

frodrigo commented on web/static/map/map.js:
```diff
@@ -37,6 +37,7 @@ export function initMap() {
   urlVars.level = urlVars.level || Cookies.get('last_level') || '1';
   urlVars.tags = urlVars.tags || Cookies.get('last_tags');
   urlVars.fixable = urlVars.fixable || Cookies.get('last_fixable');
+  urlVars.marker_uuid = urlVars.marker_uuid || Cookies.get('marker_uuid');
```
Remove the cookie part

gileri replied:
Done in e7e0f99.

frodrigo commented:
I just test it and the result is very confusing.

Getting the coordinates of the issue if just a call to:
```
http://osmose.openstreetmap.fr/api/0.3/issue/16145620-d1ac-7f04-4689-941eefdf1140
```

gileri commented:
>I just test it and the result is very confusing.

Can you expand a bit more on what is confusing ?

>Getting the coordinates of the issue if just a call to

What I meant by that is get the coordinates from existing markers. But I could

gileri commented:
>I just test it and the result is very confusing.

Can you expand a bit more ?

>Getting the coordinates of the issue if just a call to

What I tried to do by that is get the coordinates from existing markers. But you're right I could probably make the call to find the marker coordinates, and pass the response to `openPopup()` to avoid a second API call. I'll try that.

gileri commented:
The fix was in fact real simple ! What do you think of 2cc086345c0f4dd8f6bf2133715081765e364185 ?

frodrigo commented:
On the map when I click on a marker, the popup open and close immediately.

gileri commented:
> On the map when I click on a marker, the popup open and close immediately.

[I can't reproduce it on Firefox and Chrome](https://cloud.linuxw.info/s/padiGj73Fe5DCzL), even with network throttling, clicking rapidly, reloading the page, panning/zooming. Can you provide more details ? Do you have errors in the browser console ?

I also checked that the JS file being built by `webpack` is the one loaded by the browser.

frodrigo commented:
There is no error, and I have the same with or without docker, and with Firefox and Chromium.
![Peek 10-01-2021 21-45](https://user-images.githubusercontent.com/1785486/104134857-8792be00-538c-11eb-98b4-701fa94cfb2d.gif)

gileri commented:
After more than an hour of test on various scenarios I still can't reproduce it. There isn't much code triggering popup closure.
The only significant difference I see with this PR is that the popup is placed at the latlon answered by the API, so if the API answer different latlon from the marker it could lead to this. But I don't see why it would do that.

If you can, please share some additional info to help me debug that.

* Are all markers affected ? What was the steps to get in that state ?
* XHR queries done when clicking (OSM API ? Osmose API ?)
* Osmose API response of an affected marker
* the hash of the webpack output (I believe it should match between our machines) : `curl -s localhost:20009/en/map/ | grep webpack.bundle`

Thank you !

frodrigo commented:
I found the issue. 

- works http://beta.osmose.openstreetmap.fr/fr/map/#zoom=15&lat=43.71384&lon=7.34115&item=xxxx&level=1%2C2%2C3&tags=&fixable=
- does not work http://beta.osmose.openstreetmap.fr/fr/map/#zoom=15&lat=43.71384&lon=367.34116&item=xxxx&level=1%2C2%2C3&tags=&fixable=

Note the longitude. It is valid in leaflet, and the osmose backed support it.

frodrigo commented:
An other issue with the popup anchor

Original
![imagen](https://user-images.githubusercontent.com/1785486/104234803-4665e100-5454-11eb-8455-639232dbb01b.png)

PR
![imagen](https://user-images.githubusercontent.com/1785486/104234820-4b2a9500-5454-11eb-8e6f-4db72386c791.png)

gileri commented:
> I found the issue.
> Note the longitude. It is valid in leaflet, and the osmose backed support it.

Oh, I hadn't though of panning past overflowing, nice catch !

[The OSM geometry display isn't working either on master](http://osmose.openstreetmap.fr/fr/map/#item=9001&zoom=18&lat=45.187872&lon=365.73018&level=1%2C2%2C3&tags=&fixable=) (try clicking an issue, the geometry will be displayed off-viewport). Side note : [openstreetmap.org](https://www.openstreetmap.org) do wrap coordinates, maybe to thwart this exact issue ?

I'll try to fix both issues.

>An other issue with the popup anchor

On `master` the issue wasn't anchored on the marker/issue, it was anchored on the click position. An example of opening the same issue :

![Clicking the top of the marker](https://user-images.githubusercontent.com/493438/104245619-e0358a00-5464-11eb-8b76-c736d89f564d.png)
![Clicking the bottom of the marker](https://user-images.githubusercontent.com/493438/104245621-e166b700-5464-11eb-9e51-6da2c65b1178.png)

Now it's anchored to the issue coordinates, so I don't think it's that troublesome. I'll try to add a bit of padding to put the popup on top.

gileri commented:
I think I've addressed all remaining issues :
* The popup is now opened at the issue coordinates + x * 180 depending on the user's viewport
* The popup arrow is placed on the center of the marker logo, it's prettier than ever !

>The OSM geometry display isn't working either on master

I didn't find an easy solution for that, I think it should be tackled in another issue/PR.

frodrigo commented:
Ok. Look goo to me now.

Just one point, sorry, I change my mind on the name of the option. I think now it should be `issue_uuid` in place of `marker_uuid`.

gileri commented:
>Just one point, sorry, I change my mind on the name of the option. I think now it should be issue_uuid in place of marker_uuid.

No worries, changed in f973684

frodrigo commented on web/views/error/index.tpl:
```diff
@@ -46,7 +46,7 @@
 <tr><td>item</td><td><a target="_blank" href="../errors/?item={{marker['item']}}">{{marker['item']}}</a></td></tr>
 <tr><td>class</td><td><a target="_blank" href="../errors/?item={{marker['item']}}&amp;class={{marker['class']}}">{{marker['class']}}</a></td></tr>
-<tr><td>lat lon</td><td><a target="_blank" href="../map/?item={{marker['item']}}&amp;zoom=17&amp;lat={{marker['lat']}}&amp;lon={{marker['lon']}}">{{marker['lat']}}&nbsp;{{marker['lon']}}</a></td></tr>
+<tr><td>lat lon</td><td><a target="_blank" href="../map/?item={{marker['item']}}&amp;zoom=17&amp;lat={{marker['lat']}}&amp;lon={{marker['lon']}}&amp;issue_uuid={{marker['uuid']}}">{{marker['lat']}}&nbsp;{{marker['lon']}}</a></td></tr>
```
```suggestion
<tr><td>lat lon</td><td><a target="_blank" href="../map/?item={{marker['item']}}&amp;zoom=17&amp;lat={{marker['lat']}}&amp;lon={{marker['lon']}}&amp;issue_uuid={{uuid}}">{{marker['lat']}}&nbsp;{{marker['lon']}}</a></td></tr>
```

gileri replied:
Why not, done in 724ed4e.

frodrigo commented:
Links from this page also give me self closing popup.
http://localhost:20009/fr/errors/?item=3161&class=2

gileri commented:
>Links from this page also give me self closing popup

Are you sure the popup closes or is displayed elsewhere ? For me they aren't loaded or displayed at all. You can verify this by looking at the network queries in the browser debugger, no calls are made. [When clicking on an issue, it's still `marker_uuid` that's appended to the URL fragment, not `issue_uuid`.](http://beta.osmose.openstreetmap.fr/fr/map/#source=22438&item=3161&class=2&zoom=17&lat=35.886794&lon=139.463549&level=3&tags=&fixable=&issue_uuid=49440e5f-e379-69dd-6547-b698a30eea73&marker_uuid=49440e5f-e379-69dd-6547-b698a30eea73)

Looks like the python code was deployed, but not the JS. There are still references to `marker_uuid` in `web/static/map/Osmose.Marker.js` (visible in both minimized and mapped files) that don't exist after f973684. Can you check that the JS was correctly deployed ? Thank you :)

frodrigo commented:
Sorry, I remade the deploy on beta.

gileri commented:
It works now, thank you.

frodrigo commented:
> It works now, thank you.

You mean, now you see the popup close itself?

gileri commented:
I can reproduce it on beta now, and have imported Ecuador locally to account for negative coordinates.

I don't see a proper way to support infinite panning/coordinates for both OSM data and popup, so I suggest the following (dbf1128) : moving the user back to the normalized coordinates when clicking on a marker. It solves both the OSM data and popup issues, but incurs tiles loads. Do you think that's an acceptable trade-off ?

If anyone sees a better way to fix those issues please do share !

frodrigo commented:
I think we should better try to enable this option `worldCopyJump`
https://leafletjs.com/reference-1.7.1.html#map-worldcopyjump

gileri commented:
>I think we should better try to enable this option worldCopyJump

Indeed, that make the map work like [osm.org](https://github.com/osm-fr/osmose-frontend/pull/291#issuecomment-758267804). It solves both the OSM geometry and popup placement issues.

I've also fixed the stray OSM geometry after popup closure.

frodrigo commented:
`worldCopyJump` is working. But the popup still autoclose.
Branch deployed on http://beta.osmose.openstreetmap.fr/

gileri commented:
A few minutes ago an old version was still deployed, but now both popups and `worldCopyJump` are working as current code has been deployed on http://beta.osmose.openstreetmap.fr/

frodrigo commented:
> A few minutes ago an old version was still deployed, but now both popups and `worldCopyJump` are working as current code has been deployed on http://beta.osmose.openstreetmap.fr/

Ho, yes, there a cache on beta. It is ok for me on beta. Nevertheless I still have the issue on my dev setup. I will recheck it latter.

frodrigo commented:
Found issue on my side. It is now ok for me.

frodrigo commented:
Thank you for all the time you pass on this.
It is now deployed.

gileri commented:
Thank you too for the help and review ! Glad to see it deployed, that'll help me (and others I hope) fix issues.
